### PR TITLE
Update OS-model format and fix model cache bug

### PR DESCRIPTION
**Pull request by Roxeena**
- This branch solves the issue of the program sometimes crashing when loading the cache of a 3D model.
- Updates for the OS-model format to be up to date with our current cache.
- Some previous versions of the OS-model format will still be backward compatible. This is to avoid breaking models that users have made themselves in this format and to avoid that we need to update many of your models on the data server.

### CODE REVIEW

File: src/io/model/modelreaderbinary.cpp [hunk 1]
```diff
@@ -174,43 +207,69 @@ std::unique_ptr<modelgeometry::ModelGeometry> ModelReaderBinary::loadModel(
...
-                );
+
+                // @TODO: malej 2024-07-18 Temporary hack, need to update all osmodel
+                // models on the data server
+                auto vertexSize = sizeof(io::ModelMesh::Vertex);
+                if (version < VertexColorUpdateVersion) {
+                    vertexSize -= sizeof(GLfloat[3]);
+                }
+
+                fileStream.read(reinterpret_cast<char*>(&vertex), vertexSize);
```
Comment (alexanderbock) [inline src/io/model/modelreaderbinary.cpp:253]:
```suggestion
                if (version < VertexColorUpdateVersion) {
                    fileStream.read(
                        reinterpret_cast<char*>(&vertex),
                        sizeof(io::ModelMesh::Vertex) - 3 * sizeof(GLfloat)
                    );
                }
                else {
                    fileStream.read(
                        reinterpret_cast<char*>(&vertex),
                        sizeof(io::ModelMesh::Vertex)
                    );
                }
```
How about this?

File: src/io/model/modelreaderbinary.cpp [hunk 2]
```diff
@@ -221,14 +280,27 @@ std::unique_ptr<modelgeometry::ModelGeometry> ModelReaderBinary::loadModel(
...
             }
             std::vector<io::ModelMesh::Texture> textureArray;
             textureArray.reserve(nTextures);
 
             for (int32_t t = 0; t < nTextures; ++t) {
                 io::ModelMesh::Texture texture;
 
+                if (version >= SkipMarkerUpdateVersion) {
+                    // Skip marker
+                    int8_t skip;
```
Comment (alexanderbock) [inline src/io/model/modelreaderbinary.cpp:297]:
```suggestion
                    int8_t skip = 0;
```

File: src/io/model/modelreaderbinary.cpp [hunk 3]
```diff
@@ -238,21 +310,22 @@ std::unique_ptr<modelgeometry::ModelGeometry> ModelReaderBinary::loadModel(
                 texture.hasTexture = (h == 1);
 
                 // color
-                fileStream.read(reinterpret_cast<char*>(&texture.color.r), sizeof(float));
-                fileStream.read(reinterpret_cast<char*>(&texture.color.g), sizeof(float));
-                fileStream.read(reinterpret_cast<char*>(&texture.color.b), sizeof(float));
-                if (version == CurrentModelVersion) {
+                fileStream.read(
+                    reinterpret_cast<char*>(&texture.color.r),
```
Comment (alexanderbock) [inline src/io/model/modelreaderbinary.cpp:314]:
```suggestion
                    reinterpret_cast<char*>(glm::value_ptr(texture.color)),
```

File: src/io/model/modelreaderbinary.cpp [hunk 4]
```diff
@@ -292,25 +364,20 @@ std::unique_ptr<modelgeometry::ModelGeometry> ModelReaderBinary::loadModel(
...
         // Transform
-        std::array<GLfloat, 16> rawTransform;
-        fileStream.read(
-            reinterpret_cast<char*>(rawTransform.data()),
-            16 * sizeof(GLfloat)
-        );
-        glm::mat4x4 transform = glm::make_mat4(rawTransform.data());
+        GLfloat rawTransform[16];
+        fileStream.read(reinterpret_cast<char*>(rawTransform), 16 * sizeof(GLfloat));
+        glm::mat4x4 transform = glm::make_mat4(rawTransform);
```
Comment (alexanderbock) [inline src/io/model/modelreaderbinary.cpp:375]:
```suggestion
        const glm::mat4x4 transform = glm::make_mat4(rawTransform);
```

File: src/io/model/modelreaderbinary.cpp [hunk 5]
```diff
@@ -406,19 +481,18 @@ std::unique_ptr<modelgeometry::ModelGeometry> ModelReaderBinary::loadModel(
...
-                fileStream.read(reinterpret_cast<char*>(&rotZ), sizeof(float));
-                rotKeyframe.rotation = glm::quat(rotW, rotX, rotY, rotZ);
+                struct {
+                    float w = 0.f;
+                    float x = 0.f;
+                    float y = 0.f;
+                    float z = 0.f;
+                } rot;
+                fileStream.read(reinterpret_cast<char*>(&rot), 4 * sizeof(float));
+                rotKeyframe.rotation = glm::quat(rot.w, rot.x, rot.y, rot.z);
```
Comment (alexanderbock) [inline src/io/model/modelreaderbinary.cpp:495]:
```suggestion
                float rot[4];
                fileStream.read(reinterpret_cast<char*>(rot), 4 * sizeof(float));
                rotKeyframe.rotation = glm::quat(rot[0], rot[1], rot[2], rot[3]);
```

**Review by alexanderbock — APPROVED** [review]

File: src/io/model/modelreaderbinary.cpp (hunk 1 above)
Comment (Roxeena) [reply src/io/model/modelreaderbinary.cpp:253]: I prefer the other way to have less repetitive code, but I will add a comment that explains in more detail what is done and why.

File: src/io/model/modelreaderbinary.cpp [hunk 6]
```diff
@@ -174,43 +207,69 @@ std::unique_ptr<modelgeometry::ModelGeometry> ModelReaderBinary::loadModel(
...
             vertexArray.reserve(nVertices);
 
             for (int32_t v = 0; v < nVertices; ++v) {
                 io::ModelMesh::Vertex vertex;
-                fileStream.read(
-                    reinterpret_cast<char*>(&vertex),
-                    sizeof(io::ModelMesh::Vertex) - sizeof(GLfloat[3]) // @TODO: malej 2024-07-18 Temporary hack, need to update osmodel format + all models
-                );
+
+                // @TODO: malej 2024-07-18 Temporary hack, need to update all osmodel
```
[the remark this answers is not in the document]
Comment (Roxeena) [reply src/io/model/modelreaderbinary.cpp]: This is the correct behaviour, I updated the comments

**Pull request merged**